### PR TITLE
8293472: Incorrect container resource limit detection if manual cgroup fs mounts present

### PR DESCRIPTION
**Pull request by jmtd**
This is a backport of 8293472 to 8u for cgroups v2 support. Not clean: log_debug→tty->print_cr and Files.writeString → Files.write() needed, as well as adjustments for the lack of [8266490](https://bugs.openjdk.org/browse/JDK-8266490) (PID controller support) in 8u.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293472](https://bugs.openjdk.org/browse/JDK-8293472): Incorrect container resource limit detection if manual cgroup fs mounts present

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev pull/216/head:pull/216` \
`$ git checkout pull/216`

Update a local copy of the PR: \
`$ git checkout pull/216` \
`$ git pull https://git.openjdk.org/jdk8u-dev pull/216/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 216`

View PR using the GUI difftool: \
`$ git pr show -t 216`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/216.diff">https://git.openjdk.org/jdk8u-dev/pull/216.diff</a>

</details>

### CODE REVIEW

File: hotspot/test/runtime/containers/docker/DockerBasicTest.java [hunk 1]
```diff
@@ -75,5 +76,18 @@ private static void testHelloDocker() throws Exception {
...
+        DockerRunOptions opts =
+            new DockerRunOptions(imageNameAndTag, "/jdk/bin/java", "-version")
+            .addDockerOpts("-v", "/sys/fs/cgroup:/cgroups-in:ro");
+
+        // Duplicated cgroup mounts should be handled by the container detection
+        // code and should not cause any error/warning output.
+        DockerTestUtils.dockerRunJava(opts)
+            .shouldHaveExitValue(0)
+            .shouldNotMatch("\\[os,container *\\]");
+    }
```
Comment (jerboaa) [inline hotspot/test/runtime/containers/docker/DockerBasicTest.java:91]: This test has no meaning in OpenJDK 8u. There, all container related printouts are guarded with `PrintContainerInfo` which is false for this test. It makes sense for JDKs with unified logging, as there a warning was being printed to stdout/err whether or not logging was otherwise turned on or not. It's conceivable that it would catch the assert, but that would also happen in one of the cpu/memory tests with additional cgroup fs mounts.

TLDR: I suggest to skip these test changes in the backport. All it does is adding to exec time, but would always be tautologically true (especially with the `os,container` match).

File: hotspot/test/runtime/containers/cgroup/CgroupSubsystemFactory.java [hunk 2]
```diff
@@ -362,13 +407,21 @@ public static void main(String[] args) throws Exception {
...
-            test.testCgroupv1MultipleCpusetMounts(wb, test.cgroupv1MntInfoDoubleCpuset);
-            test.testCgroupv1MultipleCpusetMounts(wb, test.cgroupv1MntInfoDoubleCpuset2);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoubleCpuset);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoubleCpuset2);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoubleMemory);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoubleMemory2);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoubleCpu);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoubleCpu2);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoublePids);
+            test.testCgroupv1MultipleControllerMounts(wb, test.cgroupv1MntInfoDoublePids2);
```
Comment (jerboaa) [inline hotspot/test/runtime/containers/cgroup/CgroupSubsystemFactory.java:424]: The PIDs patch is not in 8u, but I guess this does no harm. OK.

**Review by jerboaa — COMMENTED** [review]
Mostly good. Some comments on test code. Thanks!

**Review by jerboaa — APPROVED** [review]
Looks good to me now.

**Comment by jerboaa** [comment]
This bug has all approvals now, please integrate.
